### PR TITLE
[DIR-1292] Sync between JSON Input and Form Input

### PR DESCRIPTION
**Pull request by julia-zielke**
## Description

- If you switch between the tabs JSON Input and Form Input the data is synchronized.
- Changes in the currently active tab will override the data from the other tab. If you switch to the Form Input, every additional input data from the JSON that can't be linked to a form field, will be lost.

- Run Workflow means: **What you see is what you get**
If there are additional values in the JSON Input, which are not required or even relevant for the workflow, they will be sent as well when you run the workflow. 

E2E Tests:
I added these tests in explorer/workflow/run.spec.ts

7 - Input is synced from Form Input to JSON Input
8 - Input is synced from JSON Input to Form Input
9 - Input is synced and then changed again in the current view

## Checklist

- [x] Documentation updated if required
- [x] Test coverage is appropriate
      
## Checklist Internal

- [x] Linear issue linked (e.g. [DIR-XXXX] pull request title)
- [x] Has the PR been labeled

### CODE REVIEW

**Review by stefan-kracht — CHANGES_REQUESTED** [review]
Good work. Also thanks for correcting a few typos of mine :) 

- I added some comments in the tests
- and as we have just discussed, this is my suggestion with some minor changes (please note that there is one TODO: comment). Feel free to change anything of that suggestion as you like.

```javascript
import {
  DialogClose,
  DialogFooter,
  DialogHeader,
  DialogTitle,
} from "~/design/Dialog";
import { ElementRef, useEffect, useRef, useState } from "react";
import { Tabs, TabsContent, TabsList, TabsTrigger } from "~/design/Tabs";
import { getValidationSchemaFromYaml, workflowInputSchema } from "./utils";

import Button from "~/design/Button";
import { Card } from "~/design/Card";
import Editor from "~/design/Editor";
import FormInputHint from "./FormInputHint";
import { JSONSchemaForm } from "~/design/JSONschemaForm";
import { Play } from "lucide-react";
import { ScrollArea } from "~/design/ScrollArea";
import { decode } from "js-base64";
import { pages } from "~/util/router/pages";
import { prettifyJsonString } from "~/util/helpers";
import { useFile } from "~/api/files/query/file";
import { useForm } from "react-hook-form";
import { useNavigate } from "react-router-dom";
import { useRunWorkflow } from "~/api/tree/mutate/runWorkflow";
import { useTheme } from "~/util/store/theme";
import { useToast } from "~/design/Toast";
import { useTranslation } from "react-i18next";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";

type FormInput = {
  payload: string;
};

type JSONSchemaFormSubmit = Parameters<typeof JSONSchemaForm>[0]["onSubmit"];

// TODO: move this into "~/util/helpers"
const isObject = (input: unknown): input is object =>
  z.object({}).safeParse(input).success;

const defaultEmptyJson = "{\n    \n}";

const RunWorkflow = ({ path }: { path: string }) => {
  const { toast } = useToast();
  const { t } = useTranslation();
  const theme = useTheme();
  const navigate = useNavigate();
  const { data } = useFile({ path });
  const submitButtonRef = useRef<HTMLButtonElement>(null);
  const jsonSchemaFormRef = useRef<ElementRef<typeof JSONSchemaForm>>(null);
  const validationSchema =
    data?.type === "workflow"
      ? getValidationSchemaFromYaml(decode(data?.data ?? ""))
      : null;

  // tab handling
  const isFormAvailable = validationSchema !== null;
  const tabs = ["json", "form"] as const;
  const [activeTab, setActiveTab] = useState<(typeof tabs)[number]>(
    isFormAvailable ? "form" : "json"
  );

  const [jsonInput, setJsonInput] = useState(defaultEmptyJson);
  const [formInput, setFormInput] = useState({});

  // it is possible that no data (or stale cache data) is available when this component mounts
  // and the initial value of activeTab is out of sync with the actual isFormAvailable value
  useEffect(() => {
    setActiveTab(isFormAvailable ? "form" : "json");
  }, [isFormAvailable]);

  const {
    setValue,
    formState: { isValid },
  } = useForm<FormInput>({
    defaultValues: {
      payload: defaultEmptyJson,
    },
    resolver: zodResolver(z.object({ payload: workflowInputSchema })),
  });

  const { mutate: runWorkflow, isLoading } = useRunWorkflow({
    onSuccess: ({ namespace, instance }) => {
      navigate(pages.instances.createHref({ namespace, instance }));
    },
    onError: (error) => {
      toast({
        title: t("api.generic.error"),
        description:
          error ??
          t("pages.explorer.tree.workflow.runWorkflow.genericRunError"),
        variant: "error",
      });
    },
  });

  const runButtonOnClick = () => {
    // if this workflow supports a JSON form and the json form
    // tab is active we need to trigger this form via a ref to an
    // invisible submit button (this should be optimized but a ref
    // to the form did not work)
    if (isFormAvailable && activeTab === "form") {
      // this will implicitly trigger the JSONschema forms onSubmit callback
      submitButtonRef.current?.click();
    }

    if (activeTab === "json") {
      runWorkflow({
        path,
        payload: jsonInput,
      });
    }
  };

  const jsonSchemaFormSubmit: JSONSchemaFormSubmit = (form) => {
    runWorkflow({ path, payload: JSON.stringify(form.formData) });
  };

  const syncInputData = (selectedTab: "form" | "json") => {
    if (selectedTab === "json") {
      const formState = jsonSchemaFormRef.current?.state.formData as unknown;
      const formDataObj = isObject(formState) ? formState : {};
      const formIsEmpty = Object.keys(formDataObj).length === 0;
      const formDataString = prettifyJsonString(JSON.stringify(formDataObj));
      setJsonInput(formIsEmpty ? defaultEmptyJson : formDataString);
    }

    if (selectedTab === "form") {
      let jsonInputObj: object;
      try {
        jsonInputObj = JSON.parse(jsonInput);
      } catch (e) {
        jsonInputObj = {};
      }
      setFormInput(jsonInputObj);
    }
  };

  const disableSubmit = !isValid;

  return (
    <>
      <DialogHeader>
        <DialogTitle>
          <Play /> {t("pages.explorer.tree.workflow.runWorkflow.title")}
        </DialogTitle>
      </DialogHeader>
      <div
        className="my-3 flex flex-col gap-y-5"
        data-testid="run-workflow-dialog"
      >
        <Tabs
          value={activeTab}
          onValueChange={(value) => {
            const tabValueParsed = z.enum(tabs).safeParse(value);
            if (tabValueParsed.success) {
              setActiveTab(tabValueParsed.data);
              if (isFormAvailable) syncInputData(tabValueParsed.data);
            }
          }}
        >
          <TabsList variant="boxed">
            <TabsTrigger
              variant="boxed"
              value={tabs[0]}
              data-testid="run-workflow-json-tab-btn"
            >
              {t("pages.explorer.tree.workflow.runWorkflow.jsonInput")}
            </TabsTrigger>
            <TabsTrigger
              variant="boxed"
              value={tabs[1]}
              data-testid="run-workflow-form-tab-btn"
            >
              {t("pages.explorer.tree.workflow.runWorkflow.formInput")}
            </TabsTrigger>
          </TabsList>
          <TabsContent value={tabs[0]} asChild>
            <Card
              className="h-96 w-full p-4 sm:h-[500px]"
              noShadow
              background="weight-1"
              data-testid="run-workflow-editor"
            >
              <Editor
                value={jsonInput}
                onMount={(editor) => {
                  editor.focus();
                  if (jsonInput === defaultEmptyJson) {
                    editor.setPosition({ lineNumber: 2, column: 5 });
                  }
                }}
                onChange={(newData) => {
                  if (newData != undefined) setJsonInput(newData);

                  if (typeof newData === "string") {
                    setValue("payload", newData, {
                      shouldValidate: true,
                    });
                  }
                }}
                language="json"
                theme={theme ?? undefined}
              />
            </Card>
          </TabsContent>
          <TabsContent value={tabs[1]} asChild>
            <Card className="h-96 w-full p-4 sm:h-[500px]">
              {isFormAvailable ? (
                <ScrollArea className="h-full">
                  <JSONSchemaForm
                    formData={formInput}
                    ref={jsonSchemaFormRef}
                    schema={validationSchema}
                    omitExtraData={true}
                    action="submit"
                    onSubmit={jsonSchemaFormSubmit}
                  >
                    <Button
                      type="submit"
                      ref={submitButtonRef}
                      className="hidden"
                    />
                  </JSONSchemaForm>
                </ScrollArea>
              ) : (
                <FormInputHint />
              )}
            </Card>
          </TabsContent>
        </Tabs>
      </div>
      <DialogFooter>
        <DialogClose asChild>
          <Button variant="ghost" data-testid="run-workflow-cancel-btn">
            {t("pages.explorer.tree.workflow.runWorkflow.cancelBtn")}
          </Button>
        </DialogClose>
        <Button
          type="submit"
          disabled={disableSubmit}
          loading={isLoading}
          onClick={runButtonOnClick}
          data-testid="run-workflow-submit-btn"
        >
          {!isLoading && <Play />}
          {t("pages.explorer.tree.workflow.runWorkflow.runBtn")}
        </Button>
      </DialogFooter>
    </>
  );
};

export default RunWorkflow;



```

File: ui/e2e/explorer/workflow/run.spec.ts [hunk 1]
```diff
@@ -426,3 +427,336 @@ test("it is possible to provide the input via generated form and resolve form er
...
+  await expect(
+    inputResponseAsJson,
+    "workflow was triggered and the input is the same as initially set in the other tab"
+  ).toEqual(expectedJson);
+
+  // the input in the workflow is the same like entered in the form
+  expect(
+    await page.locator(".view-line > span"),
+    "the Input was set as expected"
+  ).toContainText(["Marty", "McFly"]);
```
[the remark this answers is not in the document]
Comment (julia-zielke) [reply ui/e2e/explorer/workflow/run.spec.ts:520]: Yes, exactly. Thanks for the tip.

**Comment by julia-zielke** [comment]
@stefan-kracht 
Thank you for the feedback. I really liked your suggestions to improve the code. 👍 
 
I had a really hard time improving the End to End Tests and hope you might have an idea about this:
To extract the JSON Input from the Editor we need to loop through several span elements and add everything back together as an object to compare it to the expectedJSON. 
I think the version 2 I tried is already close to be working, but the object has the wrong formatting and needs to be trimmed with some regex.

Please check the parts of the code that I marked with "TODO" and surrounded with // *** code improvement here ***//
Thanks!

**Comment by stefan-kracht** [comment]
The solution is pretty simple if you know it :D 

```typescript
await expect(page.getByTestId("the selector to the editor")).toContainText("the expected text", {
  useInnerText: true, // <- this one will use the inner text only. So it basically strips all html 
});
```

It took us a while to find that solution in the first place. A good starting point for that kind of stuff is to look up other tests, that might also have to solve a similar problem.

I also changed one playwright selector to make it a little bit more robust against html markup changes.

Let me know if you need help with the other open open suggestions.

**Review by stefan-kracht — APPROVED** [review]

**Pull request merged**